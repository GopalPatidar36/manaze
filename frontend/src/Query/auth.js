import { gql } from "@apollo/client";

export const LOGIN = gql`
  query Login($userEmail: String!, $password: String!) {
    login(userEmail: $userEmail, password: $password) {
      token
      user {
        uid
        userEmail
        password
        firstName
        lastName
      }
    }
  }
`;
