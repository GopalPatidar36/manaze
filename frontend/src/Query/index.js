export * from "./tickets";
export * from "./user";