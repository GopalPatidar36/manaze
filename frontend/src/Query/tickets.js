import { gql } from "@apollo/client";

const CORE_FIELDS = gql`
  fragment CoreFields on Ticket {
    id
    title
    description
    priority
    status
  }
`;

const TicketList = gql`
  ${CORE_FIELDS}
  fragment ticketList on TicketList {
    count
    rows {
      ...CoreFields
    }
  }
`;

export const GET_CURRENT_USER_TICKET = gql`
  ${TicketList}
  query CurrentUserTicket($title: String, $description: String, $priority: String, $status: String, $limit: Int, $offset: Int, $order: String) {
    currentUserTicket(title: $title, description: $description, priority: $priority, status: $status, limit: $limit, offset: $offset, order: $order) {
      ...ticketList
    }
  }
`;

export const GET_ALL_TICKET = gql`
  query TicketList($title: String, $description: String, $priority: String, $status: String, $limit: Int, $offset: Int, $order: String) {
    ticketList(title: $title, description: $description, priority: $priority, status: $status, limit: $limit, offset: $offset, order: $order) {
      ...ticketList
    }
  }
  ${TicketList}
`;

export const UPDATE_TICKET = gql`
  mutation UpdateTicket($id: Int!, $title: String, $description: String, $priority: String, $status: String) {
    updateTicket(id: $id, title: $title, description: $description, priority: $priority, status: $status) {
      id
    }
  }
`;
