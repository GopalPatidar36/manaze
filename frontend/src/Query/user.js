import { gql } from "@apollo/client";

const CORE_FIELDS = gql`
  fragment CoreFields on userFields {
    uid
    firstName
    lastName
    role
    createdAt
    updatedAt
    userEmail
  }
`;

const TicketList = gql`
  ${CORE_FIELDS}
  fragment userList on UserSearchResult {
    count
    rows {
      ...CoreFields
    }
  }
`;

export const GET_CURRENT_USER = gql`
  ${CORE_FIELDS}
  query Me {
    me {
      ...CoreFields
    }
  }
`;

export const UPDATE_USER = gql`
  mutation UpdateUser($uid: String!, $firstName: String, $lastName: String) {
    updateUser(uid: $uid, firstName: $firstName, lastName: $lastName) {
      id
    }
  }
`;
