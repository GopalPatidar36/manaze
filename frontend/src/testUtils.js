// src/testUtils.js
import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { Provider } from "react-redux";
import { MockedProvider } from "@apollo/client/testing";
import { BrowserRouter as Router } from "react-router-dom";
import store from "./store";

function render(ui, { preloadedState, apolloMocks, ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <MockedProvider mocks={apolloMocks} addTypename={false}>
          <Router>{children}</Router>
        </MockedProvider>
      </Provider>
    );
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from "@testing-library/react";
// override render method
export { render };
