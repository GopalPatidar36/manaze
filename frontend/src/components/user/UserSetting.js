import React, { useEffect, useState } from "react";
import { useMutation, useQuery } from "@apollo/client";
import { GET_CURRENT_USER, UPDATE_USER } from "../../Query/index";
import { alertMessage } from "../ToastifyAlert";

const UserSetting = (props) => {
  const { loading, error, data } = useQuery(GET_CURRENT_USER);
  const user = !loading ? data?.me : {};

  const [updateUser] = useMutation(UPDATE_USER, {
    refetchQueries: [{ query: GET_CURRENT_USER }],
    onCompleted: () => alertMessage("userUpdated"),
  });

  const [isEdit, setIsEdit] = useState(true);
  const [firstName, setFirstName] = useState();
  const [lastName, setLastName] = useState();

  const handleSubmit = (e) => {
    setIsEdit((item) => !item);
    if (user.uid) {
      updateUser({ variables: { uid: user.uid, firstName, lastName } });
    }
  };

  useEffect(() => {
    if (data?.me) {
      setFirstName(data?.me.firstName);
      setLastName(data?.me.lastName);
    }
  }, [data]);

  return (
    <div className="userSetting">
      <div className="innerUserSetting">
        <h3>Personal Information</h3>
        {isEdit ? (
          <button type="submit" onClick={() => setIsEdit((item) => !item)}>
            Edit
          </button>
        ) : (
          <button type="button" onClick={() => handleSubmit()}>
            Save
          </button>
        )}
        <div className="formUserSetting">
          <div className="fieldContainer">
            <label>First Name</label>
            {!isEdit ? (
              <input value={firstName} type="text" placeholder="Enter your first name here" onChange={(ev) => setFirstName(ev.target.value)} />
            ) : (
              <p>{firstName}</p>
            )}
          </div>

          <div className="fieldContainer">
            <label>Last Name</label>
            {!isEdit ? (
              <input value={lastName} type="text" placeholder="Enter your last name here" onChange={(ev) => setLastName(ev.target.value)} />
            ) : (
              <p>{lastName}</p>
            )}
          </div>

          <div className="fieldContainer">
            <label>Role</label>
            <p>{user.role}</p>
          </div>

          <div className="fieldContainer">
            <label>User Email</label>
            <p>{user.userEmail}</p>
          </div>
        </div>
      </div>
    </div>
  );
};

export default UserSetting;
