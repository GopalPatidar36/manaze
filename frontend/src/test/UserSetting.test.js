// src/App.test.js
// import axios from "../plugin/axios";
import React from "react";
import { screen, waitFor, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { render } from "../testUtils";
import UserSetting from "../components/user/UserSetting";
import { GET_CURRENT_USER, UPDATE_USER } from "../Query/index";

jest.mock("../plugin/axios");

const mocks = [
  {
    request: {
      query: GET_CURRENT_USER,
    },
    result: {
      data: {
        me: {
          __typename: "userFields",
          id: 11,
          uid: "236139b2-e7a5-49b8-9a56-f1b71cb12653",
          userEmail: "ramp@gmail.com",
          firstName: "ram",
          lastName: "Patidar",
          role: "member",
          createdAt: "2024-07-20T12:00:00Z",
          updatedAt: "2024-07-20T12:30:00Z",
        },
      },
    },
  },
  // {
  //   request: {
  //     query: UPDATE_USER,
  //   },
  //   result: {
  //     data: {
  //       updateUser: {
  //         id: 11,
  //       },
  //     },
  //   },
  // },
];

const errorMocks = [
  {
    request: {
      query: GET_CURRENT_USER,
    },
    error: new Error("Network Error"),
  },
];

// const data = { data: "12345" };
// axios.get.mockResolvedValue({ data });
describe("fetches and displays data", () => {
  beforeEach(async () => {
    render(<UserSetting />, { apolloMocks: mocks });
  });

  test("displays personal information header", async () => {
    await waitFor(() => {
      expect(screen.getByText(/Personal Information/)).toBeInTheDocument();
    });
  });

  test("displays initial user data", async () => {
    await waitFor(() => {
      expect(screen.getByText("ram")).toBeInTheDocument();
      expect(screen.getByText("ramp@gmail.com")).toBeInTheDocument();
    });
  });

  test("updates and displays new user data", async () => {
    await waitFor(() => {
      expect(screen.getByText("ram")).toBeInTheDocument();
    });

    fireEvent.click(screen.getByText("Edit"));

    await waitFor(() => {
      fireEvent.change(screen.getByDisplayValue("ram"), { target: { value: "Gopal" } });
    });

    fireEvent.click(screen.getByText("Save"));

    await waitFor(() => {
      expect(screen.getByText("Gopal")).toBeInTheDocument();
    });
  });
});

// test("Update and Fetch data", async () => {
//   render(<UserSetting />, { apolloMocks: mocks });

//   await waitFor(() => {
//     expect(screen.getByText(/Personal Information/)).toBeInTheDocument();
//   });

//   await waitFor(() => {
//     expect(screen.getByText(/ram/)).toBeInTheDocument();
//     expect(screen.getByText(/ramp@gmail.com/)).toBeInTheDocument();
//   });
// });

// test("handles error", async () => {
//   axios.get.mockRejectedValue(new Error("Network Error"));

//   render(<App />);

//   expect(screen.getByText("Loading...")).toBeInTheDocument();

//   await waitFor(() => expect(screen.getByText("Error fetching data")).toBeInTheDocument());
// });
