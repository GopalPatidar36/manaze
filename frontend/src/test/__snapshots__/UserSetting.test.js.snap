// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches and displays data should match dom snapshot 1`] = `
<DocumentFragment>
  <div
    class="userSetting"
  >
    <div
      class="innerUserSetting"
    >
      <h3>
        Personal Information
      </h3>
      <button
        type="submit"
      >
        Edit
      </button>
      <div
        class="formUserSetting"
      >
        <div
          class="fieldContainer"
        >
          <label>
            First Name
          </label>
          <p />
        </div>
        <div
          class="fieldContainer"
        >
          <label>
            Last Name
          </label>
          <p />
        </div>
        <div
          class="fieldContainer"
        >
          <label>
            Role
          </label>
          <p />
        </div>
        <div
          class="fieldContainer"
        >
          <label>
            User Email
          </label>
          <p />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
